src/udev/udevadm-wait.c

```diff
@@ -367,6 +428,10 @@ int wait_main(int argc, char *argv[], void *userdata) {
         if (r < 0)
                 return log_error_errno(r, "Failed to set up device monitor: %m");
 
+        r = setup_periodic_timer(event);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set up periodic timer: %m");
+
```

Comment:
i think a more interesting optimization would be to simply hook into the uevent stream from the kernel instead of the one to from udev if we don't need to wait for initialization. i.e. use the other netlink group, the one udev itself listens on. That way we'd still wait for the right event, but not be delayed by udev.

Reply:
but we should figure out why. that sounds like a bug in the kernel? 

I mean, it appears uevents get lost on the CI sometimes, but are we even sure it's the kernel that loses them and not udev?

do we know if they are simply never generated? or maybe they are dropped because netlink buffer is full? But if the latter, isn't there some mechanism to detect such overruns, and then handle them somehow (maybe by rescanning all devices?)

Reply:
There are nothing I can say sure, but
> do we know if they are simply never generated?

At least from the log of the tests, uevents are not generated, at least udevd does not receive them.
See the logs in https://github.com/systemd/systemd/issues/24360.

> or maybe they are dropped because netlink buffer is full?

The possibility of this is quite small. udevadm-wait also increases the buffer size as same as the udevd uses.

> But if the latter, isn't there some mechanism to detect such overruns

This sounds nice, even though I guess the issue is not on the buffer size.